File: Sem1-2/CPP_Java2024/List3/CPP/src/Circle.cpp
#include <Circle.hpp>
#include <utils.hpp>

#include <stdexcept>
#include <cmath>

Circle::Circle(double radius) : Shape("Circle"), radius(radius)
{
    if(radius < 0)
    {
        throw std::invalid_argument(LOG_LOC() + "Circle radius cannot be negative, got: " + std::to_string(radius));
    }
}

double Circle::Circumference() const
{
    return 2.0 * M_PI * radius;
}

double Circle::Area() const
{
    return M_PI * radius * radius;
}

File: Sem1-2/CPP_Java2024/List3/JAVA/Circle.java
public class Circle extends Shape 
{
    protected final double radius;

    public Circle(double radius) throws IllegalArgumentException
    {
        super("Circle");
        if(radius < 0)
        {
            throw new IllegalArgumentException("Radius cannot be negative, got: " + radius);
        }
        this.radius = radius;
    }

    @Override
    public double Circumference()
    {
        return 2.0 * Math.PI * radius;
    }

    @Override
    public double Area()
    {
        return Math.PI * radius * radius;
    }
}
